File: src/entities/user_category/user_category.cpp
#include "user_category.hpp"

UserCategory::UserCategory(){}
UserCategory::UserCategory(int id) : id(id) {}
UserCategory::UserCategory(int id, std::string name, int id_hourly_rate, std::string service_validity_start,
                          std::string service_validity_end) : id(id),
                                                              name(name),
                                                              id_hourly_rate(id_hourly_rate),
                                                              service_validity_start(service_validity_start),
                                                              service_validity_end(service_validity_end)
                                                              {}

void UserCategory::setId(int id){
    this->id = id;
}

void UserCategory::setName(std::string name){
  this->name = name;
}

void UserCategory::setIdHourlyRate(int id_hourly_rate){
    this->id_hourly_rate = id_hourly_rate;
}

void UserCategory::setServiceValidityStart(std::string service_validity_start){
    this->service_validity_start = service_validity_start;
}

void UserCategory::setServiceValidityEnd(std::string service_validity_end){
    this->service_validity_end = service_validity_end;
}


int UserCategory::getId() const{
    return id;
}

std::string UserCategory::getName() const{
    return name;
}

int UserCategory::getIdHourlyRate() const{
    return id_hourly_rate;
}

std::string UserCategory::getServiceValidityStart() const{
    return service_validity_start;
}

std::string UserCategory::getServiceValidityEnd() const{
    return service_validity_end;
}


bool operator== (const UserCategory &uc1, const UserCategory &uc2)
{
    return uc1.id == uc2.id;
}

std::ostream& operator<<(std::ostream& os, const UserCategory& uc)
{
    os << "UserCategory - Id: " << uc.id << ", Name: " << uc.name << ", Id Hourly rate: " << uc.id_hourly_rate
        << ", Service validity start: " << uc.service_validity_start << ", Service validity end: " << uc.service_validity_end <<std::endl;
    return os;
}

UserCategoryException::UserCategoryException(const std::string & message):_message(message) {}
const char * UserCategoryException::what() const throw() {
      return _message.c_str();
}

File: uPark_client/entities/user_category.py
class UserCategory:
    def __init__(self, id = None, name = None, id_hourly_rate = None, service_validity_start = None, service_validity_end = None):
        self._id = id
        self._name = name
        self._id_hourly_rate = id_hourly_rate
        self._service_validity_start = service_validity_start
        self._service_validity_end = service_validity_end


    def set_id(self, id):
        self._id = id

    # def set_name(self, name):
    #     self._name = name

    def set_id_hourly_rate(self, id_hourly_rate):
        self._id_hourly_rate = id_hourly_rate

    def set_service_validity_start(self, service_validity_start):
        self._service_validity_start = service_validity_start

    def set_service_validity_end(self, service_validity_end):
        self._service_validity_end = service_validity_end


    def get_id(self):
        return self._id

    def get_name(self):
        return self._name

    def get_id_hourly_rate(self):
        return self._id_hourly_rate

    def get_service_validity_start(self):
        return self._service_validity_start

    def get_service_validity_end(self):
        return self._service_validity_end

    def __eq__(self, other):
        if isinstance(other, UserCategory):
            return (self._id == other._id) or (self._name == other._name)
        return NotImplemented

    def __str__(self):
        user_category_info = (f"User category Id: {self._id}, name: {self._name}, id hourly rate: {self._id_hourly_rate}, service validity start: {self._service_validity_start}, "
        f"service validity end: {self._service_validity_end}")
        return user_category_info
